mesa/space.py

```diff
@@ -256,29 +256,33 @@ def get_neighborhood(
         if neighborhood is not None:
             return neighborhood
 
-        coordinates: set[Coordinate] = set()
-
         x, y = pos
-        for dy in range(-radius, radius + 1):
-            for dx in range(-radius, radius + 1):
-                # Skip coordinates that are outside manhattan distance
-                if not moore and abs(dx) + abs(dy) > radius:
-                    continue
 
-                coord = (x + dx, y + dy)
+        range_x = self._compute_inbound_ranges(x, self.width, radius)
+        range_y = self._compute_inbound_ranges(y, self.height, radius)
 
-                if self.out_of_bounds(coord):
-                    # Skip if not a torus and new coords out of bounds.
-                    if not self.torus:
-                        continue
-                    coord = self.torus_adj(coord)
+        neighborhood = []
 
-                coordinates.add(coord)
+        for nx in range_x:
+            for ny in range_y:
+                if nx == x and ny == y and not include_center:
```

Comment:
Why is it faster that this happens for every loop, vs just discarding it at the end, like before?

Reply:
because a deletion from a list it's O(n) while for a set it is O(1) so I suppose it's (much) slower but I will give it a try

Reply:
but considering that the condition is evaluated n times maybe it could be the same

Reply:
now I thought that we can remove x and y from the ranges before the loop so that we have better perf

Reply:
wrong sorry otherwise we loose other points also

Reply:
The output of `list(a_set)` is deterministic, at least since 3.8 (see this [comment](https://github.com/projectmesa/mesa/issues/1052#issuecomment-888468826)). I haven't tested, but `list(a_set)` is likely faster than `sorted(a_set)`.

Reply:
OK, but at least for a `set[tuple[int, int]]` it is?

Reply:
I read that ints are deterministic, but don't know about tuples unfortunately, at the same type it seems like there is no guarantee on the stability of sets between different versions, so I think it can be risky to rely on that. Also, given the problem at end I don't think using sets would result in better performance actually (this discussion is anyway useful for example for move_to_empty behaviour anyway)  becuase iterating on sets is slower, creating sets is slower, we just have a better removal but I think that the total time would be slightly slower anyway

Reply:
the answer is...no:

try

```
from random import randint
a = set((randint(1,30), randint(1,30)) for x in range(100000))
b = set((randint(1,30), randint(1,30)) for x in range(100000))
# check if the len is 900 before equality between lists
print(len(a) == len(b) == 900)
print(list(a) == list(b))
```

Reply:
or yes maybe, this works for example

```
from random import randint
k = [(randint(1,30), randint(1,30)) for x in range(100000)]
a = set(k)
b = set(k)
print(list(a) == list(b))
```

Reply:
included the discarding at the end, slightly faster

Reply:
Tried to use a set for `neighborhood` just for seeing if it is faster, it is actually slightly slower (also for the user iterating over it will become slower)